### DISCUSSION TITLE
Building on Raspberry Pi 64 bit  ( arch64 )

### DISCUSSION THREAD

**Discussion by davidgritter** [GrandOrgue/GoOdf#14] [General]
When building on Raspberry Pi I get an error message about unknown compilier options.  This appears to be due to line 69 in the CMakeLists.txt file 
" add_compile_options(-O3 -mmmx -msse -msse2 -msse3 -mssse3 -mfpmath=sse -ffast-math) "
which in turn ends up in a build/CMakeFiles/GOODF.dir/flags.make file after running cmake.  If I remove the offending flags from the flags.make file, the program compiles and runs fine on Raspberry pi.

**Comment by larspalo** [comment]
Did you check exactly which of the options caused the problem? From very quickly looking at the GO cmake instructions I'm guessing that just removing the -mfpmath=sse could be enough, but I've not tested it myself.

On https://elinux.org/RPi_Software there's some recommendations, the relevant part that I find is:

```
The gcc compiler flags which will produce the most optimal code for the RPi are:

-Ofast -mfpu=vfp -mfloat-abi=hard -march=armv6zk -mtune=arm1176jzf-s

-Ofast may produce compile errors for some programs. In this case, -O3 or -O2 should be used instead. Note that -mcpu=arm1176jzf-s can be used in place of -march=armv6zk -mtune=arm1176jzf-s.
```

**Reply by davidgritter** [reply]
gcc for arm rejects all of the mmx and sse options, since they are x86 related.  for more information see my response to your next post

**Comment by larspalo** [comment]
@davidgritter could you test adding this instead of line 69 in the CMakeList.txt file?

```
  if(${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm") 
    add_compile_options(-O3 -mfpu=vfp -mfloat-abi=hard)
  else()
    add_compile_options(-O3 -mmmx -msse -msse2 -msse3 -mssse3 -mfpmath=sse -ffast-math)
  endif()
```

**Reply by davidgritter** [reply]
Your options for arm would be ok for a raspberry pi3b (arm V6), but I see better options for a pi 4(arm V8), which is complicated by the  OS

for 32 bit (traditional raspbian) the architecture is reported as arm  In that case I find that the options should be
-mfpu=neon-fp-armv8 -mfloat-abi=hard

For 64 bit raspbian which I use, the architecture is reported as aarch64  no specific SIMD or floating point options are needed as hardware floating point architecture is defined by the aarch64 requirements.  The compiler rejects the -mfpu and -mfloat-abi arguements.

in all cases the -O3 and -ffast-math arguments can be used. Unfortunately, at the moment I can only test with Pi 4 Raspbian 64.  With that platform -O3 and -ffast-math work without problem.

**Reply by larspalo** [reply]
As GOODF is not having critical issues with that its performance needing to be at the very top, do you think that it would be better to just remove any specific options and only have -O3 and -ffast-math for any release build (providing we test it first)?

**Reply by davidgritter** [reply]
A quick internet search leads me to believe that for 64 bit operating systems gcc will automatically select the appropriate sse hardware type, so both x86-64 and aarch64 don't need any floating point options.  However 32 bit linux x86 and 32 bit arm gcc will perform floating point calculations in software if the hardware FPU type is not specified.  I would assume almost everyone trying to use GrandOrgue or GOODF would by using 64 bit systems just to get enough addressable RAM, so my answer to your question above would be yes.   This would mean poorer performance on 32 bit x86 and arm systems, but floating point calculations are not happening that often in event driven software like GOODF.

Unfortunately, I don't have any 32 bit systems to test this on at the moment.  I could easily create a 32 bit linux using virtual box, though.  Let me know if you want me to do this.

**Reply by larspalo** [reply]
The only computer I have that still use a 32 bit OS is a very old Raspberry Pi (version 2 if I remember correctly) that I only use headless as a local test server for web related things. Other than that I only have 64 bit versions to test with.

For maximum compatibility/usability with different systems I think it perhaps would be best to remove any flags that won't work more or less across any system setup (that at least can use gcc/g++). There are ways to provide different presets to cmake that could provide different optimizations that the user could opt in for, but I think that's of less importance than many other things to make GOODF work really well as a tool.

**Comment by davidgritter** [comment]
I tested with and without the mmx/sse flags on a 32 bit linux virtual machine.   There was no significant difference in performance or executable size with or without the flags, so I think we can safely remove them.
